test/syscalls/linux/mount.cc

```diff
@@ -326,6 +326,14 @@ TEST(MountTest, MountFuseFilesystemNoDevice) {
   SKIP_IF(IsRunningOnGvisor() && !IsFUSEEnabled());
 
   auto const dir = ASSERT_NO_ERRNO_AND_VALUE(TempPath::CreateDir());
+
+  // Before kernel version 4.16-rc6, FUSE mount is protected by
+  // capable(CAP_SYS_ADMIN). After this version, it uses
+  // ns_capable(CAP_SYS_ADMIN) to protect. Before the 4.16 kernel, it was not
+  // allowed to mount fuse file systems without the global CAP_SYS_ADMIN.
```

Comment:
Don't we skip this test entirely if we don't have CAP_SYS_ADMIN from line 325? Is there a difference here based on namespaces?

Reply:
We are able to have `CAP_SYS_ADMIN` but still cannot mount FUSE before kernel 4.16.
```
vagrant@ubuntu1:~$ uname -a
Linux ubuntu1 4.15.0-109-generic #110-Ubuntu SMP Tue Jun 23 02:39:32 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
vagrant@ubuntu1:~$ docker run -it --rm --cap-add SYS_ADMIN --device /dev/fuse ubuntufuse
root@f5877962c374:/fus# ./hello /mnt
fuse: mount failed: Permission denied
```

However, if using kernel after 4.16, there is no error.
```
vagrant@arch:~$ $ uname -a
Linux arch.localdomain 5.7.10-arch1-1 #1 SMP PREEMPT Wed, 22 Jul 2020 19:57:42 +0000 x86_64 GNU/Linux
vagrant@arch:~$ docker run -it --rm --cap-add SYS_ADMIN --device /dev/fuse ubuntufuse
root@320fa68df5f4:/fus# ./hello /mnt
root@320fa68df5f4:/fus#
```

@avagin told me about the commit 4ad769f3c346ec3d458e255548dec26ca5284cf6 in Linux source code, it adds `FS_USERNS_MOUNT` flag into FUSE mount. This is introduced in `v4.16-rc6-11-g4ad769f3c346` revision.
```
commit 4ad769f3c346ec3d458e255548dec26ca5284cf6
Author: Eric W. Biederman <ebiederm@xmission.com>
Date:   Tue May 29 09:04:46 2018 -0500

    fuse: Allow fully unprivileged mounts

    Now that the fuse and the vfs work is complete.  Allow the fuse filesystem
    to be mounted by the root user in a user namespace.

    Signed-off-by: "Eric W. Biederman" <ebiederm@xmission.com>
    Signed-off-by: Miklos Szeredi <mszeredi@redhat.com>

diff --git a/fs/fuse/inode.c b/fs/fuse/inode.c
index 22c76cf8c2e3..48baa26993f3 100644
--- a/fs/fuse/inode.c
+++ b/fs/fuse/inode.c
@@ -1222,7 +1222,7 @@ static void fuse_kill_sb_anon(struct super_block *sb)
 static struct file_system_type fuse_fs_type = {
        .owner          = THIS_MODULE,
        .name           = "fuse",
-       .fs_flags       = FS_HAS_SUBTYPE,
+       .fs_flags       = FS_HAS_SUBTYPE | FS_USERNS_MOUNT,
        .mount          = fuse_mount,
        .kill_sb        = fuse_kill_sb_anon,
 };
```

Reply:
@mrahatm We check CAP_SYS_ADMIN in the current user namespace. But before 4.16, it requires the global CAP_SYS_ADMIN. In the linux source, there are two helpers ns_capable and capable.